add a function--ApplyAddAdditiveNoise

LvHang opened the pull request:
@pegahgh 
Hi Pegah,
I have already finished the function about add additive noise with the option "--add-noise".
Please check it.
Thank you for your guidance.

Hang

LvHang commented on src/xvectorbin/nnet3-xvector-signal-perturb-egs.cc:
```diff
@@ -73,16 +73,71 @@ void ApplyPerturbation(XvectorPerturbOptions opts,
+  // In the version, we ask the noise_cols >= input_cols. If mfcc, the cols are equal.
+  // If raw data, we ask the noise_cols > input_cols.
+  int32 input_rows = input_eg.NumRows(), input_cols = input_eg.NumCols();  
+  KALDI_ASSERT(noise_eg.NumCols() >= input_cols);
```
@pegahgh 
Hi, pegah.
I know the noise_eg.NumCols() == input_cols should be equal in feature domain, such as mfcc.
I let noise_eg.NumCols() > input_cols, just because I want to do something like you write in ApplyPerturbation(). It makes the dimensionality of the noise_eg will a little longer than input_eg. It maybe useful in raw-data situation.
I just want to make sure. Thanks a lot.
Hang

pegahgh replied:
For now, we can focus on MFCC domain, and if it gives us improvement, we can switch to raw waveform.
We may need to write different Perturbation class for raw waveform as we have more flexibility in raw waveform domain.

LvHang commented:
@pegahgh 
Hi, Pegah
I modify it. Please check it and review others. Thanks!
Hang

pegahgh commented on src/xvectorbin/nnet3-xvector-signal-perturb-egs.cc:
```diff
@@ -73,16 +73,71 @@ void ApplyPerturbation(XvectorPerturbOptions opts,
+// This function add the noise to the orginial signal. We should not normalize 
+// the signal level of the orginial signal. According to SNR, we rescale the noise
+// and add it. So that the perturbed signal is created. 
+void ApplyAddAdditiveNoise(const int32 &SNR,
```
Change the name to ApplyAdditiveNoise

pegahgh commented on src/xvectorbin/nnet3-xvector-signal-perturb-egs.cc:
```diff
@@ -73,16 +73,71 @@ void ApplyPerturbation(XvectorPerturbOptions opts,
+                                                  start_col_ind, input_cols));
+  // compute the energy of noise and input
+  Matrix<BaseFloat> input_energy_mat(input_rows, input_cols);
+  input_energy_mat.AddMatMatElements(1.0, input_eg, input_eg, 1.0);
```
Although input_energy_mat initialized with zero, it should be AddMatMatElements(1.0, input_eg, input_eg, 0.0),

pegahgh replied:
It is not a good idea to design the code like this. You should write this function in signal-distort.h and add-noise and snr  should be added as options to XvectorPerturbationOptions struct.
The function should be ApplyAdditiveNoise(const VectorBase<BaseFloat> input, const VectorBase<BaseFloat> noise,  BaseFloat snr, Vector<BaseFloat> *noisy_input)

In class PerturbXvectorSignal, you have applyDistortion which is a general function, which applies all type of distortions to input.
Then it applies distortions w.r.t opts_.

You need to add a function PerturbExamples(const XvectorOptions opts, const Matrix<BaseFloat> &input_egs, Matrix<BaseFloat> *perturbed_egs)
and this function called in nnet3-xvector-signal-perturb.cc and it generates object from PerturbXvectorSignal and vectorize the input and calls ApplyDistortion to apply different type of distortions on input.

LvHang commented:
@pegahgh 
Hi Pegah,
According to your suggestions, I modify the files. Maybe it also has some unsatisfying  points.
Could you give me some suggestion?
Thank you very much for your patient and guidance.
Hang

pegahgh commented on src/feat/signal-distort.cc:
```diff
@@ -72,5 +72,93 @@ void TimeStretch(const MatrixBase<BaseFloat> &input_egs,
+void PerturbXvectorSignal::ApplyAdditiveNoise(const MatrixBase<BaseFloat> &input_eg,
+                                              const Matrix<BaseFloat> &noise_eg,
+                                              const int32 &SNR,
+                                              Matrix<BaseFloat> *perturb_eg) {
```
the name should be perturbed_eg

pegahgh commented on src/feat/signal-distort.cc:
```diff
@@ -72,5 +72,93 @@ void TimeStretch(const MatrixBase<BaseFloat> &input_egs,
+    const kaldi::nnet3::NnetIo &noise_eg_io = noise_eg.io[0];
+    Matrix<BaseFloat> noise_eg_mat;
+    noise_eg_io.features.CopyToMat(&noise_eg_mat);
+    int32 SNR = opts_.snr;
```
Add these lines nnet3-perturb-egs binary.
I told you that You have to just call PerturbEgs function in nnet3-perturb-egs.cc. You should put loop for reading egs in nnet3 binary not here!
PerturbEgs should be defined as a separate function not a function of this class. The point is that in PertubEgs function, you create object from class PerturbXvectorSignal and call ApplyDistortion.

pegahgh commented on src/feat/signal-distort.h:
```diff
@@ -32,6 +32,7 @@
 #include "feat/resample.h"
 #include "matrix/matrix-functions.h"
 #include "cudamatrix/cu-matrix.h"
+#include "nnet3/nnet-example.h"
```
remove it, it is a wrong dependency!

pegahgh commented on src/feat/signal-distort.h:
```diff
@@ -70,8 +79,14 @@ class PerturbXvectorSignal {
+
+  void ApplyAdditiveNoise(const MatrixBase<BaseFloat> &input_eg,
+                          const Matrix<BaseFloat> &noise_eg,
+                          const int32 &SNR,
```
remove SNR

pegahgh commented on src/xvectorbin/nnet3-xvector-signal-perturb-egs.cc:
```diff
@@ -73,16 +73,28 @@ void ApplyPerturbation(XvectorPerturbOptions opts,
+                    const Matrix<BaseFloat> &input_egs,
+                    Matrix<BaseFloat> *perturb_egs) {
+  //new a PerturbXvectorSignal object and call ApplyDistortion
+  PerturbXvectorSignal perturb_xvector(opts);
```
no need to generate new object from opts!!

pegahgh replied:
1. Change perturb_egs to perturbed_eg
2. Change perturb_xvector  to perturb_egs.

LvHang commented:
@pegahgh 
Hi Pegah,
Accorading to our discussion, I modify the files.
I check some other option structures. In general, they will be the one-dimensional data type, such as int, double, string and so on. So I didn't add a matrix to XvectorPerturbOptions. I add a private point in class PerturbXvectorSignal. Maybe we can discuss and find a better solution.
Please check the binaries. Thank you very much for your guidance.
Hang

pegahgh commented on src/feat/signal-distort.cc:
```diff
@@ -72,5 +72,81 @@ void TimeStretch(const MatrixBase<BaseFloat> &input_egs,
+void PerturbXvectorSignal::ApplyDistortion(const MatrixBase<BaseFloat> &input_egs,
+                                           Matrix<BaseFloat> *perturb_egs) {
+    // conduct ApplyAdditiveNoise
+  if (!opts_.add_noise_rspecifier.empty()) {
```
change option name to add_noise

pegahgh replied:
I think the best strategy is to have a add-noise option  in PertubXvectorOption as noise rspecifier not noise examples.
--add-noise=noise.scp, where noise.scp corresponds to features for different noises. You can randomly select different noises.
Then you no longer need to pass noise matrix to PerturbExample and you can easily pass noise rspecifier using --add-noise option.
You don't need to change ApplyAdditiveNoise class. You just need to check if add-noise is not empty  in ApplyDistortion and the read matrix of noise using BaseFloatMatrixReader and pass it to ApplyAdditiveNoise.

LvHang commented:
@pegahgh 
Hi Pegah,
I modify it. Now, we pass a filename, such as noise.scp, to add-noise option. And we choose a noise matrix in ApplyDistortion function. I hope I understand your intention in a right way.
Please check it. Thanks a lot for your patience.
Hang

pegahgh commented on src/feat/signal-distort.cc:
```diff
@@ -72,5 +72,97 @@ void TimeStretch(const MatrixBase<BaseFloat> &input_egs,
+  if (!opts_.add_noise.empty()) {
+    // choose a noise from the noise.scp/ark
+    // 1) we need to record the keys of noise_egs
+    std::vector<std::string> list_noise_egs;
```
It is no longer noise_egs

pegahgh commented on src/feat/signal-distort.cc:
```diff
@@ -72,5 +72,97 @@ void TimeStretch(const MatrixBase<BaseFloat> &input_egs,
+    ApplyAdditiveNoise(input_egs, *noise_egs_, perturb_egs);
+    // conduct others
+    // TODO
+  } else { // deal with the opts_.noise_egs situation
```
no need for else condition!
We can compose several different perturbation to generated perturbed_egs.

pegahgh commented on src/feat/signal-distort.cc:
```diff
@@ -72,5 +72,97 @@ void TimeStretch(const MatrixBase<BaseFloat> &input_egs,
+  }
+}
+
+// This function is a entrance. It calls ApplyDistortion to apply different
```
Change the comment!

pegahgh commented on src/feat/signal-distort.h:
```diff
@@ -59,19 +63,35 @@ struct XvectorPerturbOptions {
     opts->Register("noise-egs", &noise_egs, "If supplied, the additive noise is added to input signal.");
     opts->Register("rand_distort", &rand_distort, "If true, the signal is slightly changes"
                    "using some designed FIR filter with no zeros.");
+    opts->Register("add-noise", &add_noise, "specify a file contains some noise egs");
```
change the definition! e.g. Noise rspecifier for additive noises, if nonempty, the additive noise randomly selected and added to input egs.

pegahgh commented on src/feat/signal-distort.h:
```diff
@@ -59,19 +63,35 @@ struct XvectorPerturbOptions {
   }
 };
 
 class PerturbXvectorSignal {
```
Add Comment about PerturbXvectorSignal class

pegahgh commented on src/feat/signal-distort.cc:
```diff
@@ -72,5 +72,97 @@ void TimeStretch(const MatrixBase<BaseFloat> &input_egs,
+    Matrix<BaseFloat> noise_eg_mat = noise_random_reader.Value(key_noise_eg);
+    SetNoiseEgs(noise_eg_mat);
+
+    ApplyAdditiveNoise(input_egs, *noise_egs_, perturb_egs);
```
You can directly use noise_mat, why do you use noise_egs_?

LvHang commented:
@pegahgh 
Hi Pegah,
According to your suggestions, I modified the binaries. I removed the private variable "*noise_egs_" in class PerturbXvectorSignal and other corresponding changes.
However, I haven't add comment to class PerturbXvectorSignal. I have no idea and I'm afraid I will make some mistakes. 
Please check it. Thanks a lot.
Hang

pegahgh commented on src/feat/signal-distort.h:
```diff
@@ -59,19 +63,27 @@ struct XvectorPerturbOptions {
     opts->Register("noise-egs", &noise_egs, "If supplied, the additive noise is added to input signal.");
```
remove this option

pegahgh commented on src/xvectorbin/nnet3-xvector-signal-perturb-egs.cc:
```diff
@@ -51,8 +51,8 @@ void ApplyPerturbation(XvectorPerturbOptions opts,
+    ////                              &rand_distort_shifted_egs);
   }
 
   if (noise_egs) { 
```
remove if condition

pegahgh commented on src/xvectorbin/nnet3-xvector-signal-perturb-egs.cc:
```diff
@@ -129,19 +130,26 @@ int main(int argc, char *argv[]) {
+      // add
+      if (!perturb_opts.add_noise.empty()) {
+        // We call the PerturbExample to implement adding distortion.
+        PerturbExample(perturb_opts, input_eg_mat, &perturb_eg_mat);
```
You should not call PertubExample here, it shoule be called in nnet3-perturb-egs!

pegahgh commented on src/xvectorbin/nnet3-xvector-signal-perturb-egs.cc:
```diff
@@ -129,19 +130,26 @@ int main(int argc, char *argv[]) {
+        // We call the PerturbExample to implement adding distortion.
+        PerturbExample(perturb_opts, input_eg_mat, &perturb_eg_mat);
+      } else {
+        ApplyPerturbation(perturb_opts, input_eg_mat, noise_mat, &perturb_eg_mat);
```
you should call this in PerturbExample not here

pegahgh commented on src/xvectorbin/nnet3-xvector-signal-perturb-egs.cc:
```diff
@@ -129,19 +130,26 @@ int main(int argc, char *argv[]) {
       const NnetIo &noise_io = noise_egs.io[0];
-      noise_io.features.CopyToMat(noise_mat);
-       
+      noise_io.features.CopyToMat(noise_mat);       
```
remove these lines
you don't need to read noises in this stage.

pegahgh replied:
add PerturbXvectorOptions to options for this binary.

pegahgh commented on src/xvectorbin/nnet3-xvector-signal-perturb-egs.cc:
```diff
@@ -51,8 +51,8 @@ void ApplyPerturbation(XvectorPerturbOptions opts,
-    perturb_xvector.ComputeAndApplyRandDistortion(shifted_egs,
-                                  &rand_distort_shifted_egs);
+    ////perturb_xvector.ComputeAndApplyRandDistortion(shifted_egs,
+    ////                              &rand_distort_shifted_egs);
```
remove ApplyPerturbation function from nnet3-perturb-egs. We will process egs in distort-signal.*

LvHang commented:
@pegahgh 
Hi Pegah,
I modify the main program-- "xvectorbin/nnet3-xvector-signal-perturb-egs.cc".
And remove the noise--egs option from struct XvectorPerturbOptions.
Please check it. Thank you very much.
Hang

pegahgh commented on src/feat/signal-distort.cc:
```diff
@@ -72,5 +72,90 @@ void TimeStretch(const MatrixBase<BaseFloat> &input_egs,
+      std::string key = noise_seq_reader.Key();
+      noise_list.push_back(key);
+    }
+    noise_seq_reader.Close();
```
This part is not efficient, since you need to generate list of noises, each time you call this function for each example.
The best thing is that you add private variable std::vector<std::string> noise_lists_   to PertubXvectorSignal and then initialize that in class constructor and then you can randomly pick a noise from noise list in ApplyDistortion.

Also PerturbExample should accept object of PerturbXvectorSignal as input as PerturbExample(PertubXvectorSignal &example_perturber, Matrix<BaseFloat> &eg, Matix<BaseFloat> *perturbed_eg) and you should generate object of PertubXvectorSignal class before example loop.

Another alternative is to remove ApplyDistortion function from  PertubXvectorSignal class and Do all those steps in PertubExample and then make AddAdditiveNoise as public function in Xvector class.

LvHang commented:
@pegahgh 
Hi Pegah,
I update the files depend on what we discussed this afternoon.
Please check it. Thanks a lot.
Hang

pegahgh commented:
Hang, the code looks fine. The next step is to prepare x-vector setup for our feature extraction.

LvHang commented:
@pegahgh 
Hi Pegah,
I wrote the nnet3-fvector-get-egs.cc and Makefile. Please check it whether it comply with your mind.
Thanks a lot.
Hang

pegahgh commented:
Hi Hang
Thanks, I will review your repo now.

On Tue, Dec 13, 2016 at 8:37 PM, LvHang <notifications@github.com> wrote:

> @pegahgh <https://github.com/pegahgh>
> Hi Pegah,
> I wrote the nnet3-fvector-get-egs.cc and Makefile. Please check it whether
> it comply with your mind.
> Thanks a lot.
> Hang
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pegahgh/kaldi/pull/10#issuecomment-266917059>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AI-PtOrn0zEtR0SZ9gtGnzmnkjePIgg6ks5rH0hPgaJpZM4K-piE>
> .
>

LvHang commented:
@pegahgh 
Hi Pegah,
I update "generate_noise_range.py" and "lh_add_noise.sh"
Please briefly check it and don't care the style. I think I just want to focus on the function and I will modify the style later.
Now I let the range format: wav_t_start, wav_t_end, noise_name, noise_t_start, noise_t_end, snr
I think I have a few questions and we can discuss them tomorrow.
For the binary--"perturb-signal.cc", it will depend on the range format. So I just write a version and I will modify it after our discussion.
Thanks a lot for your guidance!
Hang

LvHang commented:
@pegahgh 
Hi Pegah,
I update the files "generate-fixed-length-range.py" "generate-variable-length-range.py" and "nnet3-fvector-perturb-signal.cc" in vector version.
Please check it. If you think the function is satisfying. I will change the design.
In the original version, we use the  class PerturbXvectorSignal (feat/signal-distort.h). Now we call it fvector, so I haven't combine them.
Thank you for your guidance.
Hang

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/generate_variable_length_range.py:
```diff
@@ -0,0 +1,188 @@
+#!/usr/bin/env python
```
What is the difference between this function and generate_noise_ranges.py??

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/generate_variable_length_range.py:
```diff
@@ -0,0 +1,188 @@
+# This is the variable-length version
+# The format is <wav_t_start> <wav_t_end> <noise_uttid> <noise_t_start> <noise_t_end> <snr>
+
+# For <wav_t_start> <wav_t_end>
```
You can add option --variable-len-additive-noise and then combine two functions. 
You can have random SNR for both of them!

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/lh_add_noise.sh:
```diff
@@ -0,0 +1,41 @@
+min_additive_noise_len=2 # the minimum duration of each noise file
+num_kind_range=4         # the number of kinds of noise ranges
+min_snr=0                # the minimum snr value
+max_snr=0                # the maximum snr value
```
The default values for these variables don't make sense!

pegahgh commented on src/feat/signal-distort.cc:
```diff
@@ -72,5 +72,87 @@ void TimeStretch(const MatrixBase<BaseFloat> &input_egs,
+// This function add the noise to the orginial signal. We should not normalize 
+// the signal level of the orginial signal. According to SNR, we rescale the noise
+// and add it. So that the perturbed signal is created. 
+void PerturbXvectorSignal::ApplyAdditiveNoise(const MatrixBase<BaseFloat> &input_eg,
```
It is better to have separate function for applying noise than XvectorPerturbSignal class.
In the next step, you need to add extra binary as perturb-wav in featbin, which processes noise-range and apply it to input wave.
pertub-wav --noise=scp:noise.scp --noise-range="single line from range-file you generated" wav perturbed_wav
You can have struct NoiseRange with options in the range and Then it initialize  std::vector<NoiseRange> noise_ranges using range-line and define function ApplyNoise(std::string noise_scp, const std::vector<NoiseRange> noise_ranges, const VectorBase<BasFloat> &input_wav,  VectorBase<BaseFloat> *pertub_wav)
You need to check sampling frequency and ..
Then you need to add perturbed version of wav to wav.scp as perturb{perturb_id}-wavname

LvHang commented:
Hi Pegah,

I do some modifications. Please check it.
1. For python files. I update the generate_fixed_length_range.py and generate_variable_length_range.py according to your suggestion. And I generate a integrate version--generate_noise_range.py. In the integrate version, we use --variable-len-additive-noise (true|false) to generate the fixed/variable range-file. I have sent the specific example of each noise range file to you this afternoon. I added the wav-id-perturb-{i} in front of each line.
2. I write the binary--nnet3-fvector-perturb-signal.cc. In this file, I split the single line--"NoiseRange" and generate the controller--vector<NoiseController>. Then I use ApplyNoise() to add noise to wav with controller. You can find the details in binary.
3. I generate a simple version bash script to call the python file. In the future, it can call the binary--nnet3-fvector-perturb-signal.cc.
Thank you very much for your guidance.

Best wishes,
Hang

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/add_noise.sh:
```diff
@@ -0,0 +1,64 @@
+cmd=run.pl
+nj=4
+# Begain Configuration.
+min_additive_noise_len=2.0       # the minimum duration of each noise file
```
add unit in description (is it sec?)

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/add_noise.sh:
```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
```
add description to script.

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/add_noise.sh:
```diff
@@ -0,0 +1,64 @@
+nj=4
+# Begain Configuration.
+min_additive_noise_len=2.0       # the minimum duration of each noise file
+num_kind_range=4                 # the number of kinds of noise ranges
```
I think kind is not appropriate name, you can rename it to num_ranges_per_wav

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/generate_fixed_length_range.py:
```diff
@@ -0,0 +1,191 @@
+
+# The <snr> control the rate of signal and noise. In the other word, scale the amplitude of noise.
+# The snr will be randomly selected form the range (max-snr, min-snr).
+
```
Function description needs some modifications.

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/generate_fixed_length_range.py:
```diff
@@ -0,0 +1,191 @@
+# remove the noise whose length < --min-additive-noise-len
+num_error = 0
+num_done = 0
+f = open(args.noise2dur, "r")
```
It is better to add separate function ass wave2dur with output as wav_id and wav_len and use it for both noise2dur and wav2dur.

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/generate_fixed_length_range.py:
```diff
@@ -0,0 +1,191 @@
+        continue
+
+    # create a file to record the ranges
+    f = open(args.range_dir + "/" + str(wav_ids[i]) + ".noiserange", "w")
```
do you still generate separate noiserange file per wave?

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/generate_fixed_length_range.py:
```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/env python
+
+# The function use to generate range file for fvector
```
I think you forget to remove this file from git.

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/generate_noise_range.py:
```diff
@@ -0,0 +1,303 @@
+# The function use to generate range-file for fvector
+# Each line of the range-file corrsponds to a kind of perturbed wav. In each line,
+# we use comma to seperate different addnoise range. The format of each addnoise
+# range is <wav_t_start>:<wav_t_end>:<noise_uttid>:<noise_t_start>:<noise_t_end>:<snr>
```
add perturbed-wave-id in the beginning of range line.

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/generate_noise_range.py:
```diff
@@ -0,0 +1,303 @@
+	sys.exit("ERROR: The wav length is shorter than --min-additive-noise-len")
+    wav_ids.append(a[0])
+    wav_lengths.append(float(a[1]))
+f.close()
```
It is better to add separate function ass wave2dur with output as wav_id and wav_len and use it for both noise2dur and wav2dur.

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/generate_noise_range.py:
```diff
@@ -0,0 +1,303 @@
+    num_fixed_error = 0
+    num_fixed_done = 0
+
+    for i in range(0, wav_num_utts):
```
wav_num_utts is not appropriate name!

pegahgh replied:
num_wav is better name.

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/generate_noise_range.py:
```diff
@@ -38,90 +76,228 @@
+                current_snr = random.randrange(args.max_snr, args.min_snr)
 
-            print("{0} {1} {2} {3} {4} {5} {6}".format(wav_t_start,
+                print("{0}:{1}:{2}:{3}:{4}:{5}".format(wav_t_start,
```
you still don't add pertubed-wav-id in the begining?

pegahgh commented on src/fvectorbin/nnet3-fvector-perturb-signal.cc:
```diff
@@ -36,6 +36,78 @@ struct NoiseController{
     noise_t_start_(noise_t_start), noise_t_end_(noise_t_end), snr_(snr) { }
 };
 
+void GenerateController(std::vector<std::string> &segments, 
```
NoiseContoller is not a appropriate name! PertubWavRange

pegahgh commented on src/fvectorbin/nnet3-fvector-perturb-signal.cc:
```diff
@@ -36,6 +36,78 @@ struct NoiseController{
+  // add noise
+
+  for (int i=0; i < controller.size(); ++i) {
+    const WaveData &noise_wav = noise_reader.Value(controller[i].noise_uttid_);
```
variable in struct should not have "_" at end, just class variable has "_". Also it should be noise_id not noise_uttid!!

pegahgh commented on src/fvectorbin/nnet3-fvector-perturb-signal.cc:
```diff
@@ -36,6 +36,78 @@ struct NoiseController{
+                                    input_end_point - input_start_point + 1);
+    SubVector<BaseFloat> noise_part(noise, noise_start_point,
+                                    noise_end_point - noise_start_point + 1);
+    Vector<BaseFloat> selected_noise(input_part.Dim());
```
add description for this part.

pegahgh commented on src/fvectorbin/nnet3-fvector-perturb-signal.cc:
```diff
@@ -36,6 +36,78 @@ struct NoiseController{
+      while (the_rest > noise_part.Dim()) {
+        selected_noise.Range(selected_noise.Dim()-the_rest,
+                             noise_part.Dim()).CopyFromVec(noise_part);
+        the_rest = the_rest - noise_part.Dim();
```
the_rest is not appropriate name.

pegahgh commented on src/fvectorbin/nnet3-fvector-perturb-signal.cc:
```diff
@@ -44,27 +116,27 @@ int main(int argc, char *argv[]) {
-        "nnet3-fvector-perturb-signal --noise-range-file=uttid.range.n --add-noise-list="
-        "scp:noise.scp --input-channel=0 input.wav output.wav\n";
+        "nnet3-fvector-perturb-signal --noise=scp:noise.scp --noise-range="
+        "\"head -n 5 a.noiserange | tail -n 1\" --input-channel=0 input.wav "
```
don't write head -n ... in the usage. Write actual values in range line for description.

pegahgh commented on src/fvectorbin/nnet3-fvector-perturb-signal.cc:
```diff
@@ -77,23 +149,15 @@ int main(int argc, char *argv[]) {
-        controller.push_back(NoiseController(wav_t_start, wav_t_end, noise_uttid,
-                                            noise_t_start, noise_t_end, snr));
-      }
+    if (noise_range != "") {
```
it (!noise_range.empty())

pegahgh commented on src/fvectorbin/nnet3-fvector-perturb-signal.cc:
```diff
@@ -116,52 +180,9 @@ int main(int argc, char *argv[]) {
-      BaseFloat scale_factor = sqrt(input_energy/ noise_energy/ (pow(10, snr/20)) );
-      output.Range(input_start_point, input_part.Dim()).AddVec(scale_factor, selected_noise);
-    }
+    ApplyNoise(noise, controller, input, &output);
```
Rename name "controller in all files.

pegahgh commented:
1. I reviewed your codes and added comments
1.1. You should remove unnecessary files from git.
1.2. you should add pertubed-wav-id in the beginning of range line. 
1.3. You should write all ranges for all waves in a same file as ranges.

2. In the next step, you need to generate two files using python file. 
2.1. wav2pertubedwav , it is similar to spk2utt 
wav1  wav1-p1 wav1-p2 wav1-p3
In future, it can contains different perturbed version of wav like reverberated wav or combined spks.
It is better to generate this file in generate-noise-ranges.py.

Then you can combine wav2pertubedwav genereted using different perturbation function in future.

2.2. You should use ranges.scp and wav2perturbedwav and data/train/wav.scp as input to python file  generate-pertub-wav-specifier.py, and it will generate new wav.scp e.g. data/train_pertubed/wav.scp 
wav.scp -> it is the same as old wav.scp, but it has perturbed version of wav added to file
e.g.
wav1 sph2pipe -f wav -p -c 1 $path/wav1.sph |
wav1-p1 sph2pipe -f wav -p -c 1 $path/wav1.sph | nnet3-fvector-perturb-signal --noise-scp=scp:noise.scp noise-range="range-p1-for-wav1" - |
wav1-p2 sph2pipe -f wav -p -c 1 $path/wav1.sph | nnet3-fvector-perturb-signal --noise-scp=scp:noise.scp noise-range="range-p2-for-wav1" - |
...

LvHang commented:
@pegahgh 
Hi Pegah,
According your suggestions, I fixed the generate_noise_range.py, add_noise.sh and nnet3-fvector-perturb-signal.cc.
Please check it.
I will submit the files about step2 later tonight.

Thanks
Hang

LvHang commented:
@pegahgh 
Hi Pegah,
According to your suggestions, I modified the generate_noise_range.py and add_noise.sh. Now
1.1 I remove unnecessary files in github
1.2 I add pertubed-wav-id in the beginning of range line and all ranges for all waves in the same file. At the same time, I generate wav2pertubedwav in generate_noise_range.py.
1.3 I write the file--generate-pertub-wav-specifier.py which could generate the perturbed_wav.scp as you show.
But, in generate-pertub-wav-specifier.py, there is a problem. When <recording-id> != <utterance-id>, it will make mistake.
The reason is : The general wav.scp format is <recording-id> <extended-filename>. However, when the segment file(<utt-id> <recording-id> <begin> <end>) is exist, the utt2dur is <utt-id> <duration>. And our range is generated from utt2dur. So we use the <utt-id>. I will different from the <recording-id> so that make the mistake. 
When the data directory doesn't have segment, <recording-id> == <utt-id>, the python code works well.
I think may be we need a redirection with segment. I will try it tomorrow.
Please check it. Thank you for your guidance.

Hang

LvHang commented:
Hi Pegah,
I modify the files in steps/nnet3/fvector directory.
1. If the segments is exist, I will temporary remove it so that the format of file-utt2dur is <recording-id> <duration>, and I will use the new utt2dur file to produce the perturbed_wav.scp. Thus, it will avoid the problem that "<utt-id> != <recording-id>"
2. I convert "wav1 wav1-p1 wav1-p2 wav1-p3" to "wav1 p1-wav1 p2-wav1 p3-wav1". I refer to the old kaldi scripts, it always has the prefix other than suffix.

I will deal with update the utt2spk and so on in data directory. The examples have been sent to your mailbox.

Hang

LvHang commented:
@pegahgh 
Hi Pegah,
I write a apply_map_one2mult.pl to complete the mapping form 1 to multiple. It will help in generate utt2spk, segments, text, spk2utt and reco2file_and_channel.
I modify the add_noise.sh. Now it will generate a integrated directory such as data/perturb/ which contains all the  efficient files.

Best wishes,
Hang

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/generate_perturb_wav_specifier.py:
```diff
@@ -0,0 +1,98 @@
+            wav_extended_files[current_wav_index],
+            args.noise,
+            perturbed_range_contents[current_perturbed_wav_index]),file=g)
+g.close()
```
All variable name you used is not informative, but we can fix it later.

pegahgh commented:
I expect the code to compile and run properly. 
The next step is to generate MFCC features and examples for perturbed data.

1. You can easily generate mfcc features using modified wav.scp, so you need to generate separate data dir as a copy of original one with new wav.scp generated using generate-perturb-wav-specifier.py.
also utt2spk and spk2utt needs to modify using new modified utterances.
you can also use wav2perturbedwav and spk2utt to generate utt2pertubedutt.
orig-utt1 utt1-p1 utt1-p2 utt1-p3  
2. In the next step, you can use utt2perturbedutt to generate egs(it is the same as allocate_example.py in xvector)
e.g.
utt{i}-p{j} utt{i}-p{k}  <relative-archive-index> <abs-archive-index> <start-frame-index> <num-frames>

LvHang commented:
Hi Pegah,
1. I fix the bug about segments. Now add_noise.sh can generate the correct segments file.
2. I think the file--"perturb_utt_map" can be treat as utt2pertubedutt
2. According to xvector/allocate_examples.py, I wrote the fvector/allocate_examples.py. 
2.1. I want to make sure one thing. In fvector, assume we have M archives, each archive contains N examples and each example has two chunks. The lengths of two chunks are len1 and len2 separately. I know len1 == len2 and the length of chunk in certain archive is same. I want to know whether the length of chunk in different archives are the same. (In my python code, all chunk length is a specified value). Is it right?
2.2 For each utterance, we have ori_utt and pertubed{i}_utt. So I think the chunk pair can be 'ori_utt{i} p{j}-utt{i}' or 'p{j}-utt{i} p{k}-utt{i}', right?
3. I'm writing the nnet3-fvector-get-egs which uses to resolve the output of allocate_examples.py. I will finished it today. 
Please check it. And what can we do next step?

Best wishes,
Hang

LvHang commented:
Hi Pegah,
I complete the C++ code about nnet3-fvector-get-egs.cc which use to deal with the output of allocate_examples and generate the egs. Please check it.
I will write the shell to connect all the steps: generate perturbed data directory; make mfcc; get egs. And test them.
Maybe we can go to the next step about the simple experiments.

Best wishes,
Hang

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/allocate_examples.py:
```diff
@@ -0,0 +1,234 @@
+    f.close()
+
+    f = open(args.egs_dir + "/temp/" + prefix + "outputs." + str(job + 1), "w")
+    if f is None:
```
I don't really like temp for directory name. you can rename it as aux_info.

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/allocate_examples.py:
```diff
@@ -0,0 +1,234 @@
+        utt_len = lengths[utt_index]
+        offset = GetRandomOffsets(utt_len, args.frames_per_chunk)
+        this_egs.append( (utt_index, offset) )
+    all_egs.append(this_egs)
```
In some situations, the variable all_egs gets large cause some io problem and it may slow down the program.
So it maybe better to write down and read egs for different archives. What is the running speed for allocate_egs in your case?
Also do you discard the utterances, that you generated pairs for them?

LvHang replied:
From line 180 to 189:
I generate a 'egs list' for each archive. Each item in the 'egs list' is a tuple (utt, offset). 
In later code, I will use each tuple to generate the final information which will be parsed by nnet3-fvector-get-egs.cc. 
I will use 'utt' to randomly select two utterances from '<ori-utt-id> <p1-utt-id> ... <pn-utt-id>' to generate '<source-utterance1> <source-utterance2>'. And 'offset' is used to assign "<offset>", which means the distance between the beginning of utterance and the start frame.
So, at last, I will generate the information as followed:
<source-utterance1> <source-utterance2> <relative-archive-index> <absolute-archive-index> <offset> <frame-length>

I'm not sure the running speed. I just use a few utterance to test the code is right or not. I use the same method with David to do this part.

pegahgh commented on egs/wsj/s5/steps/nnet3/fvector/allocate_examples.py:
```diff
@@ -0,0 +1,234 @@
+    this_egs = [ ] # this will be an array of 2-tuples (utterance-index, start-frame).
+    for n in range(this_num_egs):
+        utt_index = RandomUttAtLeastThisLong(args.frames_per_chunk)
+        utt_len = lengths[utt_index]
```
you should have a variable exclude_utt and add utt_index to exclude_utt and then just choose random utterance using (all_utt - excluded_utt) for each archive.

pegahgh commented on src/fvectorbin/nnet3-fvector-get-egs.cc:
```diff
@@ -0,0 +1,211 @@
+
+      std::string utt_a = fields[0],
+                  utt_b = fields[1],
+                  start_frame_str = fields[4],
```
fields[2] and fields[3]. You just copied from David's code without changing the name!!

LvHang replied:
What do you mean?
The form of the information is:
<source-utterance1> <source-utterance2> <relative-archive-index> <absolute-archive-index> <offset> <frame-length>
So fields[2] and fields[3] are same with David's '<relative-archive-index> <absolute-archive-index>'.

pegahgh commented:
@LvHang 
The setup is almost ready for doing some preliminary experiments. You can start doing some preliminary experiment using this setup.
You can start doing experiment on small dataset (e.g. 20% swbd ~ 60hrs) for now and generate 5-10 perturbed version of data using noise.scp, (for now you can use 20% of speakers 
Do you have any sample egs generated using this setup?
The next step for generating noise is as follows:
1. Use music and point-source noises in /export/a09/pegahgh/kaldi-xvector/egs/aspire/s5/RIR_NOISES.
to generated perturbed data using wsj/s5/steps/nnet3/fvector/generate_perturb_wav_specifier.py.
(I think you need to generate noise.scp)

2. In the next step, use egs/aspire/s5/local/multi_condition/reverberate_data_dir.sh to generate reverberated wavs using RIR_NOISES.
3.  combine data-dirs from (1) and (2) and start generating egs for fvector experiment.
Be careful to have at most 10 perturbed version of each waveform.
You need to generate features for all of them and generate egs for fvector.
Let me know when you prepared egs for fvector experiments. 
 

Meanwhile, we can start preparing our raw-waveform setup for fvector(I think all codes for this experiment is also ready. am I right??

LvHang commented:
@pegahgh 
1. Firstly, I will use small data set, part of swbd corpus, to test my codes and perfect them
2. Then, I will generate the egs using RIR_NOISE and reverberated RIR_NOISE.
3. I think you are right. We can preparing our raw-wavform setup. During the period, we can find what we lack and supplement it.

LvHang commented:
Hi Pegah,
I have finished the test. Now we can use steps/nnet3/fvector/add_noise.sh to complete the process of generating the egs.
I use music noises in  /export/a09/pegahgh/kaldi-xvector/egs/aspire/s5/RIR_NOISES(use sox down sampling to 8kHz) to generated perturbed the swbd/data/train. The perturbed version is 7, the min-snr=-10, the max-snr=-20. You can get the perturbed data dir in /export/a11/hlyu/kaldi-now/egs/swbd/s5c/data/perturbed_music. And the egs in /export/a11/hlyu/kaldi-now/egs/swbd/s5c/exp/fvector_a/perturbed_music/egs.

Two problems:
1. For the point-source noises, they are always 1 second, but the average length of swbd file is 200 seconds. That means we need about 200 noise files to cover a swbd recording.Do we really have to do this?
2. For reverberated noises, which is impulse response dir？we use it to reverberate the swbd data and treat the reverberated swbd as the noise? Is my understanding right?

Best wishes 
Hang
